pkg/api/controller_test.go

```diff
@@ -1150,6 +1197,133 @@ func TestController_DeleteBranchHandler(t *testing.T) {
+		require.NotNil(t, resp.JSON201.Range)
+		require.NotNil(t, resp.JSON201.Pagination)
+		require.Equal(t, count, resp.JSON201.Range.Count)
+		require.Equal(t, strings.Replace(w.Entries[0].FullKey, uriPrefix, prepend, 1), resp.JSON201.Range.MinKey)
+		require.Equal(t, strings.Replace(w.Entries[count-1].FullKey, uriPrefix, prepend, 1), resp.JSON201.Range.MaxKey)
+		require.False(t, resp.JSON201.Pagination.HasMore)
+		require.Empty(t, resp.JSON201.Pagination.LastKey)
+		require.Empty(t, resp.JSON201.Pagination.ContinuationToken)
+	})
+
+	t.Run("successful ingestion with pagination", func(t *testing.T) {
+		// force splitting the range before
+		count := 200_000
+		clt, w := setup(count, nil)
+
+		resp, err := clt.IngestRangeWithResponse(ctx, "repo1", api.IngestRangeJSONRequestBody{
+			After:             after,
+			FromSourceURI:     fromSourceURIWithPrefix,
+			Prepend:           prepend,
+			ContinuationToken: &token,
+		})
+
+		verifyResponseOK(t, resp, err)
+		require.NotNil(t, resp.JSON201.Range)
+		require.NotNil(t, resp.JSON201.Pagination)
+		require.Less(t, resp.JSON201.Range.Count, count)
+		require.Equal(t, strings.Replace(w.Entries[0].FullKey, uriPrefix, prepend, 1), resp.JSON201.Range.MinKey)
+		require.Equal(t, strings.Replace(w.Entries[resp.JSON201.Range.Count-1].FullKey, uriPrefix, prepend, 1), resp.JSON201.Range.MaxKey)
+		require.True(t, resp.JSON201.Pagination.HasMore)
+		require.Equal(t, w.Entries[resp.JSON201.Range.Count-1].FullKey, resp.JSON201.Pagination.LastKey)
+		require.Equal(t, testutils.ContinuationTokenOpaque, *resp.JSON201.Pagination.ContinuationToken)
+	})
+
+	t.Run("error during walk", func(t *testing.T) {
+		// force splitting the range before
+		count := 10
+		expectedErr := errors.New("failed reading for object store")
+		clt, _ := setup(count, expectedErr)
+
+		resp, err := clt.IngestRangeWithResponse(ctx, "repo1", api.IngestRangeJSONRequestBody{
+			After:             after,
+			FromSourceURI:     fromSourceURIWithPrefix,
+			Prepend:           prepend,
+			ContinuationToken: &token,
+		})
+
+		require.NoError(t, err)
+		require.Equal(t, http.StatusInternalServerError, resp.StatusCode())
+		require.Contains(t, string(resp.Body), expectedErr.Error())
+	})
+}
+
+func TestController_WriteMetaRangeHandler(t *testing.T) {
+	ctx := context.Background()
+	clt, deps := setupClientWithAdmin(t)
+
+	// setup test data
+	_, err := deps.catalog.CreateRepository(ctx, "repo1", onBlock(deps, "foo1"), "main")
+	testutil.Must(t, err)
+
+	t.Run("successful metarange creation", func(t *testing.T) {
+		resp, err := clt.CreateMetaRangeWithResponse(ctx, "repo1", api.CreateMetaRangeJSONRequestBody{
+			Ranges: []api.RangeMetadata{
+				{Count: 11355, EstimatedSize: 123465897, Id: "FirstRangeID", MaxKey: "1", MinKey: "2"},
+				{Count: 13123, EstimatedSize: 123465897, Id: "SecondRangeID", MaxKey: "3", MinKey: "4"},
+				{Count: 10123, EstimatedSize: 123465897, Id: "SecondRangeID", MaxKey: "5", MinKey: "6"},
```

Comment:
Question - in real life, the fact that we will get two ranges with the same ID for metarange is almost not possible, right?

Reply:
Right, please see the second bullet in the PR description. We currently don't validate it.